src/geometry/quaternion_construction.rs

```diff
@@ -676,6 +676,54 @@ impl<N: RealField> UnitQuaternion<N> {
     where SB: Storage<N, U3> {
         Self::new_eps(axisangle, eps)
     }
+
+    /// Create the mean unit quaternion from a vector of unit quaternions.
+    ///
+    /// Algorithm from: Oshman, Yaakov, and Avishy Carmi. "Attitude estimation from vector
+    /// observations using a genetic-algorithm-embedded quaternion particle filter." Journal of
+    /// Guidance, Control, and Dynamics 29.4 (2006): 879-891.
+    /// 
+    /// # Example
+    /// ```
+    /// # #[macro_use] extern crate approx;
+    /// # use std::f32;
+    /// # use nalgebra::{UnitQuaternion};
+    /// let q1 = UnitQuaternion::from_euler_angles(0.0, 0.0, 0.0);
+    /// let q2 = UnitQuaternion::from_euler_angles(-0.1, 0.0, 0.0);
+    /// let q3 = UnitQuaternion::from_euler_angles(0.1, 0.0, 0.0);
+    ///
+    /// let quat_vec = vec![q1, q2, q3];
+    /// let q_mean = UnitQuaternion::quaternions_mean(&quat_vec);
+    ///
+    /// let euler_angles_mean = q_mean.euler_angles();
+    /// assert_relative_eq!(euler_angles_mean.0, 0.0, epsilon = 1.0e-7)
+    /// ```
+    #[inline]
+    pub fn quaternions_mean(unit_quaternions: &Vec<Self>) -> Self {
```

Comment:
```suggestion
    pub fn mean_of(unit_quaternions: &[Self]) -> Self {
```

Or maybe even

```suggestion
    pub fn mean_of(unit_quaternions: impl IntoIterator<Item=Self>) -> Self {
```

Reply:
In your first suggestion you changed the vector argument into a slice. However, both of them are used as a reference. What would be the advantage of using a slice?

For your second suggestion, would it allow any structure that implement IntoIterator? If yes it is pretty neat! But apparently the compiler is complaining about the `map` function.

>  = note: the method `map` exists but the following trait bounds were not satisfied:
            `&mut impl IntoIterator<Item = Self> : std::iter::Iterator`
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `map`, perhaps you need to implement one of them:
            candidate #1: `std::iter::Iterator`
            candidate #2: `generic_array::functional::FunctionalSequence`

 I suppose I need to specify that the elements we are iterating on are mutable. Could you please guide me on how to write it (I almost never used traits in Rust yet) ?

Reply:
> What would be the advantage of using a slice?

Things other than vectors can produce a slice; for example, statically sized stack-allocated arrays.

> But apparently the compiler is complaining about the map function.

`map` is a method on `Iterator`, not `IntoIterator`. You need to call `into_iter` to get an `Iterator` from an `IntoIterator`.